### PR TITLE
Object Type-Detection and Replacing object.assign

### PR DESCRIPTION
**Pull request by misteroneill**
## Description
This pulls in a few updates I made to `utils/obj` from #3690 so they can be more isolated and make it into 5.x. 

Perhaps more impactful, this removes `object.assign` as a dependency and re-implements the functionality we were depending on. _This change reduces the minified size of Video.js by over 2KB off the gzipped size._

## Specific Changes proposed
- Introduce the `isObject` method: essentially, `typeof x === 'object'` with a null check.
- Introduce the `isPlain` method: taken from `mergeOptions`, this does a reasonable job at identifying plain objects.
- Remove `object.assign`.

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

File: src/js/utils/obj.js [hunk 1]
```diff
@@ -30,3 +30,52 @@ export function reduce(object, fn, initial = 0) {
...
+ * Returns whether a value is an object of any kind - including DOM nodes,
+ * arrays, regular expressions, etc. Not functions, though.
+ *
+ * This avoids gotchas like `typeof null === 'object'`.
+ *
+ * @param  {Object} value
+ * @return {Boolean}
+ */
+export function isObject(value) {
+  return !!value && typeof value === 'object';
```
Comment (gkatsev) [inline src/js/utils/obj.js]: this directly contradicts the jsdoc comment.

File: src/js/utils/obj.js [hunk 2]
```diff
@@ -30,3 +30,52 @@ export function reduce(object, fn, initial = 0) {
...
+
+/**
+ * Returns whether an object appears to be a "plain" object - that is, a
+ * direct instance of `Object`.
+ *
+ * @param  {Object} value
+ * @return {Boolean}
+ */
+export function isPlain(value) {
+  return isObject(value) && value.constructor === Object;
```
Comment (gkatsev) [inline src/js/utils/obj.js]: why remove the `toString` check.

**Comment by BrandonOCasey** [comment]
is there any reason not to use the `mergeOptions` in place of `assign` rather than having a new function? Is there any reason to even have `mergeOptions`? should `mergeOptions` just use this new function?

**Comment by misteroneill** [comment]
@BrandonOCasey Yeah, I do think we ought to not have the partial lodash dependency there, but that function is a bit more complex in that it's recursive and has the "customizer" thing; so, it's a bit more prone to unforeseen breaking changes. That said, I do plan to take a stab at it.

File: src/js/utils/obj.js (hunk 2 above)
Comment (misteroneill) [inline src/js/utils/obj.js]: I think a copy-paste mistake.

File: src/js/utils/obj.js (hunk 1 above)
Comment (misteroneill) [inline src/js/utils/obj.js]: Will clarify! 👍

**Comment by BrandonOCasey** [comment]
SPACE SAVING AHOY! 🚀

File: src/js/utils/fn.js [hunk 3]
```diff
@@ -31,7 +31,34 @@ export const bind = function(context, fn, uid) {
...
+ *
+ * @param  {Function} fn
+ *         The function to be throttled.
+ *
+ * @param  {Number}   wait
+ *         The number of milliseconds by which to throttle.
+ *
+ * @return {Function}
+ */
+export const throttle = function(fn, wait) {
```
Comment (gkatsev) [inline src/js/utils/fn.js]: can this be done in a separate PR?

File: src/js/utils/obj.js [hunk 4]
```diff
@@ -30,3 +31,55 @@ export function reduce(object, fn, initial = 0) {
...
+}
+
+/**
+ * Returns whether an object appears to be a "plain" object - that is, a
+ * direct instance of `Object`.
+ *
+ * @param  {Object} value
+ * @return {Boolean}
+ */
+export function isPlain(value) {
```
Comment (BrandonOCasey) [inline src/js/utils/obj.js]: would it be better to use https://lodash.com/docs/4.16.6#isPlainObject?

File: src/js/utils/obj.js [hunk 5]
```diff
@@ -30,3 +31,55 @@ export function reduce(object, fn, initial = 0) {
...
+ * Returns whether a value is an object of any kind - including DOM nodes,
+ * arrays, regular expressions, etc. Not functions, though.
+ *
+ * This avoids the gotcha where using `typeof` on a `null` value
+ * results in `'object'`.
+ *
+ * @param  {Object} value
+ * @return {Boolean}
+ */
+export function isObject(value) {
```
Comment (BrandonOCasey) [inline src/js/utils/obj.js]: would it be better to just use https://lodash.com/docs/4.16.6#isObject?

Comment (misteroneill) [inline src/js/utils/obj.js]: My next PR is eliminating lodash as a dependency. It's not worth the extra weight of including lodash just for this.
